src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java

```diff
@@ -812,8 +819,12 @@ KindSelector attribArgs(KindSelector initialKind, List<JCExpression> trees, Env<
      *  Check that all the types are references.
      */
     List<Type> attribTypes(List<JCExpression> trees, Env<AttrContext> env) {
+        return attribTypes(trees, env, i -> false);
+    }
+
+    List<Type> attribTypes(List<JCExpression> trees, Env<AttrContext> env, Predicate<JCExpression> valueOK) {
```

Comment:
Do you really need a predicate here? Wouldn't a boolean be ok? Or perhaps a flag in env, or some other means?

Reply:
the thing is that one value only doesn't apply to all cases, what if some type variables are universal but others aren't?

Reply:
it wont be that simple, see that the method invoked is doing:
```
    List<Type> types = attribAnyTypes(trees, env);
    return chk.checkRefTypes(trees, types, valueOK);
```

and chk.checkRefTypes is doing:
```
        List<JCExpression> tl = trees;
        for (List<Type> l = types; l.nonEmpty(); l = l.tail) {
            l.head = checkRefType(tl.head.pos(), l.head, valueOK.test(tl.head));
            tl = tl.tail;
        }
        return types;
```
and Check.checkRefType is the one using the flag, so in order to do all this at the level of `Attr.visitTypeApply` I will have to unfold all this code and then the code in `Attr.visitTypeApply` will lose readability. This is why I opted for this solution.

Reply:
I see what you mean - it seems like the current code is working against us here. In fact, attribAnyTypes specifically says that clients are expected to check for reference types after the fact...

Here's an idea... you could maybe call attribAnyTypes yourself, to attribute all the types, and then, in the loop you already have, only call checkRefTypes on the types you really need?

Reply:
OK